packages/controls/src/controls/color/color.ts

```diff
@@ -0,0 +1,201 @@
+          .optional(),
+        defaultValue: resolvedValue,
+        hideAlpha: z.boolean().optional(),
+      })
+
+      const definition = z.object({
+        type,
+        config,
+        version,
+      })
+
+      return {
+        type,
+        value,
+        resolvedSwatchValue,
+        resolvedValue,
+        data,
+        config,
+        version,
+        definition,
+      }
+    }
+
+    return {
+      version,
+      relaxed: schemas(z.string().optional()),
+      strict: schemas(z.string()),
+    }
+  }
+
+  static deserialize(data: DeserializedRecord): ColorDefinition {
+    if (data.type !== Definition.type)
+      throw new Error(
+        `Color deserialization: expected '${Definition.type}', got '${data.type}'`,
+      )
+
+    const { config, version } = Definition.schema.relaxed.definition.parse(data)
+    return new ColorDefinition(config, version)
+  }
+
+  constructor(
+    config: C,
+    readonly version: z.infer<typeof Definition.schema.version>,
+  ) {
+    super(config)
+  }
+
+  get controlType() {
+    return Definition.type
+  }
+
+  get schema(): ReturnedSchemaType<C> {
+    return Definition.schema.relaxed
+  }
+
+  get dataSchema() {
+    return (
+      (this.config.defaultValue === undefined
+        ? Definition.schema.relaxed
+        : Definition.schema.strict) as Schema<C>
+    ).data
+  }
+
+  safeParse(data: unknown | undefined): ParseResult<DataType<C> | undefined> {
+    return safeParse(this.dataSchema, data)
+  }
+
+  fromData(data: DataType<C> | undefined): ValueType<C> | undefined {
+    const inputSchema = this.dataSchema.optional()
+    return match(data satisfies z.infer<typeof inputSchema>)
+      .with(Definition.dataSignature.v1, ({ swatchId, alpha }) => ({
+        swatchId,
+        alpha,
+      }))
+      .otherwise((value) => value)
+  }
+
+  toData(value: ValueType<C>): DataType<C> {
+    return match(this.version)
+      .with(1, () => ({
+        ...Definition.dataSignature.v1,
+        ...value,
+      }))
+      .with(undefined, () => value)
+      .exhaustive()
+  }
+
+  copyData(
+    data: DataType<C> | undefined,
+    { replacementContext }: CopyContext,
+  ): DataType<C> | undefined {
+    if (data == null) return data
+
+    const replaceSwatchId = (swatchId: string) =>
+      replacementContext.swatchIds.get(swatchId) ?? swatchId
+
+    const inputSchema = this.dataSchema.optional()
+    return match(data satisfies z.infer<typeof inputSchema>)
+      .with(Definition.dataSignature.v1, (val) => ({
+        ...val,
+        swatchId: replaceSwatchId(val.swatchId),
+      }))
+      .otherwise((val) => ({
+        ...val,
+        swatchId: replaceSwatchId(val.swatchId),
+      }))
+  }
+
+  createInstance(sendMessage: SendMessage<any>) {
+    return new DefaultControlInstance(sendMessage)
+  }
+
+  serialize(): [SerializedRecord, Transferable[]] {
+    return serialize(this.config, {
+      type: Definition.type,
+      version: this.version,
+    })
+  }
```

Comment:
Re: `createInstance`, not straightforwardly; `createInstance` in the `ControlDefinition` can return a derived instance type, thus `ControlDefinition` cannot provide a default implementation of `createInstance` that returns `DefaultControlInstance` since this could be a lie.

Re: `serialize`, we can provide a default implementation that serializes config + type, which would cover 80% of the controls. I choose not to because I wanted to force the question of serialization on the control author (in this case, ourselves) as we were going through all of the controls. We can definitely revise this if we feel that there is no need to force this on the authors of any new controls.